src/files.ts

```diff
@@ -187,22 +193,61 @@ export async function collectIgnoreRules(
     // Check if symlink and exclude if requested
     if (!fileStats || (fileStats.isSymbolicLink() && !symlinksEnabled) || fileStats.isFile()) return [];
 
-    // Find ignore files inside this directory
-    const localIgnoreFiles = await fg(
+    // Find .snyk and .[*]ignore files inside this directory.
+    const allIgnoredFiles = await fg(
       IGNORE_FILES_NAMES.map(i => `*${i}`),
       {
         ...fgOptions,
         cwd: folder,
         followSymbolicLinks: symlinksEnabled,
       },
     );
-    // Read ignore files and merge new patterns
-    return union(...localIgnoreFiles.map(parseFileIgnores));
+
+    // Parse rules from all .snyk files inside this directory.
+    const snykFiles = allIgnoredFiles.filter(f => f.endsWith(DOTSNYK_FILENAME));
+    const snykExcludeRules = union(...snykFiles.map(parseFileIgnores));
+
+    // Parse rules from all .[*]ignore files inside this directory.
+    const ignoreFiles = allIgnoredFiles.filter(
+      // Exclude .snyk files (parsed separately above) and files in paths excluded by .snyk files.
+      f => !f.endsWith(DOTSNYK_FILENAME) && multimatch(f, snykExcludeRules).length === 0,
+    );
+
+    return parseIgnoreRulesFromFiles(ignoreFiles, snykExcludeRules);
   });
+
   const customRules = await Promise.all(tasks);
+
   return union(fileIgnores, ...customRules);
 }
 
+/**
+ * Parse and collect ignore rules from .[*]ignore files.
+ * Removes redundant rules for efficiency.
+ * Removes negative rules that are superseded by .snyk exclude rules.
+ */
+function parseIgnoreRulesFromFiles(ignoreFiles: string[], snykExcludeRules: string[]): string[] {
+  // Read ignore files and merge patterns.
+  const localIgnoreRules = union(...ignoreFiles.map(parseFileIgnores));
+
+  const rulesAfterProcessing: string[] = [];
+  return union(localIgnoreRules, snykExcludeRules).filter((rule: string) => {
+    // Identify and remove negative rules that are superseded by .snyk exclude rules.
+    // i.e. if a directory "foo" is ignored by a .snyk exclude rule, then a negative match
+    // for "!foo" from another ignore file is superseded and needs to be removed.
+    const negativeRuleIsSupersededBySnykExcludes =
+      rule.startsWith('!') && snykExcludeRules.some(existingRule => multimatch(rule.slice(1), existingRule).length > 0);
+    // Deduplicate rules, such as "**/dir" and "sub/dir".
+    // In this case "sub/dir" is a subset of the first rule and thus redundant.
+    const ruleIsRedundant = rulesAfterProcessing.some(existingRule => multimatch(rule, existingRule).length > 0);
+
+    const ruleIsValid = !negativeRuleIsSupersededBySnykExcludes && !ruleIsRedundant;
+    if (ruleIsValid) rulesAfterProcessing.push(rule);
```

Comment:
enh: Maintaining an array of "rules we have processed" inside a filter means this entire process depends a lot on iteration order (and is a bit of an anti-pattern). A better approach would be to make this a two step process:
- Filter out the redundant rules
- Filter out the redundant rules based on the previous results

q: Is the redundancy check even necessary?

Reply:
The redundant rule filtering is not really critical and meant to reduce the number of ignore patterns we give the library, i.e. remove more specific rules if a general rule is available that would match anyway. It functionally doesn't matter if we process these our of order and might miss removing redundant rules, the more important part is filtering out those superseded by `.snyk` rules which are collected beforehand.

Reply:
This filtering was removed in the latest implementation.